### PR TITLE
Issue #10404 - Allow Users to edit inserted image

### PR DESCRIPTION
**Pull request by ivorpad**
With this feature, users will now be able to edit images when clicking on them.

Screen Shot 2020-05-08 at 7 39 53 PM

### CODE REVIEW

File: dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js [hunk 1]
```diff
@@ -2,1174 +2,1088 @@ dojo.provide("dotcms.dijit.image.ImageEditor");
...
+		this.currentUrl = this.baseFilterUrl;
+	},
+
+	createImageWindow: function () {
+		this.tabindex = 0;
+		this.thumbnailDiv.tabindex = 0;
+
+		var newerInode = window.contentAdmin.contentletInode;
+
+		console.log({ newerInode });
```
Comment (jdotcms) [inline dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js:107]: is ok to leave this here?

Comment (ivorpad) [reply dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js:107]: No, removing.

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/main/webapp/html/js/dotcms/dijit/image/image_tools.css [hunk 2]
```diff
@@ -247,7 +247,7 @@ table {font-size:inherit;font-family:Roboto,sans-serif;}
         
 #controlTable td{
 	border:0px solid silver;
-	padding:3px;	
+	padding:3px 0;
```
Comment (ivorpad) [inline dotCMS/src/main/webapp/html/js/dotcms/dijit/image/image_tools.css:250]: Added this to fix a UI issue in the 'edit image' slider.

File: dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js [hunk 3]
```diff
@@ -2,1174 +2,1088 @@ dojo.provide("dotcms.dijit.image.ImageEditor");
...
+	saveAsIncrement: 1,
+	editImageText: "Edit Image",
+	binaryFieldId: "",
+	fieldContentletId: "",
+	saveAsFileName: "",
+	compression: "none",
+	compressionValue: 65,
+	fileSize: 0,
+	execute: null,
+	activeEditor: undefined,
```
Comment (ivorpad) [inline dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js]: Unfortunately, I messed up with the formatting so I'll comment on each change made in this file.

Added two props: `execute` and `activeEditor`. The `execute()` function is used to call `createImageWindow` from  the WYSIWYG editor.

`activeEditor` is the instance of the active editor so we can insert the image in the current editor.

File: dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js [hunk 4]
```diff
@@ -2,1174 +2,1088 @@ dojo.provide("dotcms.dijit.image.ImageEditor");
...
+
+	saveImage: function () {
+		if (this.binaryFieldId != null && this.binaryFieldId.length > 0) {
+			this.saveBinaryImage(this.activeEditor);
+		} else {
+			this.saveFileImage();
+		}
+	},
+
+	_isValidURL: function (url) {
```
Comment (ivorpad) [inline dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js]: Added this helper for a quick check in `saveBinaryImage`, the reason for this is that the URL comes in either host + path or just path, so we need to count for that.

File: dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js [hunk 5]
```diff
@@ -2,1174 +2,1088 @@ dojo.provide("dotcms.dijit.image.ImageEditor");
...
+	 */
+	saveBinaryImage: function (activeEditor) {
+		var field = this.binaryFieldId;
+		if (this.fieldContentletId.length > 0) {
+			field = this.fieldContentletId;
+		}
+
+		var url = this.cleanUrl(this.currentUrl);
+
+		if (activeEditor) {
```
Comment (ivorpad) [inline dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js]: Added these lines to insert the image back to the WYSIWYG editor.

File: dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js [hunk 6]
```diff
@@ -2,1174 +2,1088 @@ dojo.provide("dotcms.dijit.image.ImageEditor");
...
-
-
-            }
-        });
-    }
-
-
-});
\ No newline at end of file
+			activeEditor.execCommand("mceReplaceContent", false, asset);
```
Comment (ivorpad) [inline dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js]: Replaced the image with the new `asset`

File: dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js [hunk 7]
```diff
@@ -384,57 +373,79 @@ dojo.declare("dotcms.dijit.image.ImageEditor", dijit._Widget,{
...
+            this.saveBinaryImage(this.activeEditor);
         }
         else{
             this.saveFileImage();
         }
     },
 
 
-
+_isValidURL: function (url) {
```
Comment (fmontes) [inline dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js:384]: Duplicated.

File: dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js [hunk 8]
```diff
@@ -384,57 +373,79 @@ dojo.declare("dotcms.dijit.image.ImageEditor", dijit._Widget,{
...
-        
-
+        var url = this.cleanUrl(this.currentUrl);
 
-        this.setThumbnail();
-        
+            if (activeEditor) {
+                var newUrl;
+                if (this._isValidURL(url)) {
+                    newUrl = new URL(url).pathname;
```
Comment (fmontes) [inline dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js:411]:
```
newUrl = this._isValidURL(url) ? new URL(url).pathname : url
```

File: dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js [hunk 9]
```diff
@@ -384,57 +373,79 @@ dojo.declare("dotcms.dijit.image.ImageEditor", dijit._Widget,{
...
+                    src="${newUrl}" 
+                    alt="${this.fieldName}"
+                    data-field-name="${this.fieldName}"
+                    data-inode="${this.inode}"
+                    data-identifier="${this.fieldContentletId}"
+                    data-saveas="${this.saveAsFileName}"
+                />`;
+            activeEditor.execCommand("mceReplaceContent", false, asset);
+		} else {
+            url = url.indexOf("?") > -1 ? url + "&" : url + "?";
```
Comment (fmontes) [inline dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js:427]: Maybe extract all this logic to a function.

File: dotCMS/src/main/webapp/html/js/tinymce/js/tinymce/plugins/doteditimage/plugin.min.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
+dojo.require("dotcms.dijit.image.ImageEditor");
+tinymce.PluginManager.add("doteditimage", function (editor) {
+	const lastSelectedImage = undefined;
+	function addButton() {
+		editor.addButton("editimage", {
+			icon: "image",
+			tooltip: "Edit Image",
```
Comment (fmontes) [inline dotCMS/src/main/webapp/html/js/tinymce/js/tinymce/plugins/doteditimage/plugin.min.js:7]: This needs translation.

**Review by fmontes — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js (hunk 9 above)
Comment (erickgonzalez) [reply dotCMS/src/main/webapp/html/js/dotcms/dijit/image/ImageEditor.js:427]: and maybe some doc for future understanding

File: dotCMS/src/main/webapp/html/js/tinymce/js/tinymce/plugins/doteditimage/langs/en.js [hunk 11]
```diff
@@ -0,0 +1,3 @@
+tinyMCE.addI18n("en", {
```
Comment (dev-dotcms) [inline dotCMS/src/main/webapp/html/js/tinymce/js/tinymce/plugins/doteditimage/langs/en.js:1]: Codacy Issue found: [Avoid trailing commas in object or array literals](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5494719)

File: dotCMS/src/main/webapp/html/js/tinymce/js/tinymce/plugins/doteditimage/langs/es.js [hunk 12]
```diff
@@ -0,0 +1,3 @@
+tinyMCE.addI18n("es", {
```
Comment (dev-dotcms) [inline dotCMS/src/main/webapp/html/js/tinymce/js/tinymce/plugins/doteditimage/langs/es.js:1]: Codacy Issue found: [Avoid trailing commas in object or array literals](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5494719)

**Review by freddyucv — APPROVED** [review]

**Pull request merged**